File: Kotlin/80.remove_duplicates_from_sorted_array_II.kt
fun main() {
    println(removeDuplicates2(intArrayOf(1,1,1,2,2,3)))
    println(removeDuplicates2(intArrayOf(0,0,1,1,1,1,2,3,3)))
    print(removeDuplicates2(intArrayOf(-3, -1)));
}

fun removeDuplicates2(nums: IntArray): Int {
    var list = mutableListOf<Int>()

    var beforePrevious:Int? = null;
    var previous:Int? = null;

    for(i in nums) {
        if(i != previous || previous != beforePrevious) {
            list.add(i);
        }

        beforePrevious = previous;
        previous = i;
    }

    list.forEachIndexed { i, v -> 
        nums[i] = v;
    }

    return list.size;
}

File: lib/80.remove_duplicates_from_sorted_array_II.dart
void main() {
  print(removeDuplicates([1, 1, 1, 2, 2, 3]));
  print(removeDuplicates([0, 0, 1, 1, 1, 1, 2, 3, 3]));
  print(removeDuplicates([-3, -1]));
}

int removeDuplicates(List<int> nums) {
  int? beforePrevious = null;
  int? previous = null;
  for (int i = nums.length - 1; i >= 0; i--) {
    if (nums[i] == previous && previous == beforePrevious) {
      nums.removeAt(i);
    }

    beforePrevious = previous;
    previous = nums[i];
  }

  return nums.length;
}
